### PR TITLE
common: Fix assertion when disabling and re-enabling clog_to_monitors

### PR DESCRIPTION
**Pull request by gerald-yang**
osd: Fix assertion when disabling and re-enabling clog_to_monitors

Steps to reproduce:

1. Disable clog_to_monitors for OSDs
ceph tell osd.* injectargs '--clog_to_monitors=true'

2. Wait for OSDs to generate some clogs

3. Re-enable clog_to_monitors
ceph tell osd.* injectargs '--clog_to_monitors=true'

This will trigger an assertion

When clog_to_monitors is disabled, "last_log" still keeps increasing by get_next_seq() if OSD writes info to clog
```c++
if (log_to_monitors) {
  e.seq = parent->queue(e);
} else {
  e.seq = parent->get_next_seq();
}
```
but "last_log_sent" doesn't increase, if we disable clog_to_monitors for a bit longer then re-enabling it
the num_unsent will be huge
unsigned num_unsent = last_log - last_log_sent;

even bigger than log_queue size and lead to the following assertion
ceph_assert(num_unsent <= log_queue.size());

This patch does the following
1. when disabling clog_to_monitors, clear log_queue
2. even logs are not sent to monitors, keep last_log_sent the same as last_log, so there won't be a lots of unsent logs when re-enabling clog_to_monitors

Fixes: https://tracker.ceph.com/issues/48946

Signed-off-by: Gerald Yang <gerald.yang@canonical.com>

### CODE REVIEW

File: src/common/LogClient.cc [hunk 1]
```diff
@@ -165,6 +165,8 @@ void LogChannel::update_config(map<string,string> &log_to_monitors,
   int graylog_port = atoi(graylog_port_str.c_str());
 
   set_log_to_monitors(to_monitors);
+  if (!to_monitors)
```
Comment (tchaikov) [inline src/common/LogClient.cc:168]: i'd suggest create two commits:

1. move `LogChannel::set_log_to_monitors(bool v)` to .cc file 
2. add a method to `LogClient`, something like `LogClient::reset()`, which clears `log_queue` and sync `last_log_sent` with `last_log`. and accompany this method with comment explaining that it avoid the hole in the disabling-then-enabling case. yes, `log_queue.clear()` would be a no-op when we re-enable `log_to_monitor`, but simpler this way.
3. in `LogChannel::set_log_to_monitors()` if `log_to_monitors` is different from the new `v`, in addition to assigning `v` to `log_to_monitors`, call `parent->reset()`.

for two reason:

1. better performance, we don't need to acquire the lock twice for getting the seq and for syncing `last_log_sent` when `log_to_monitors` is disabled
2. simpler this way.

also, please always add a space after `if`. and add curly brackets around the if block even if there is only a single statement in it, for better readability.

Comment (tchaikov) [reply src/common/LogClient.cc:168]: also, could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

**Comment by tchaikov** [comment]
@gerald-yang ping?

**Comment by gerald-yang** [comment]
@tchaikov sorry, I was too busy on other stuff in these few days, I started to work on this yesterday and should be able to update the PR today

File: src/common/LogClient.cc [hunk 2]
```diff
@@ -133,6 +133,10 @@ LogClient::LogClient(CephContext *cct, Messenger *m, MonMap *mm,
 {
 }
 
+void LogChannel::set_log_to_monitors(bool v) {
```
Comment (tchaikov) [inline src/common/LogClient.cc:136]: @@gerald-yang could you put the opening bracket "{" on a new line so it is consistent with other part of this source. also it improves the readability.

File: src/common/LogClient.cc [hunk 3]
```diff
@@ -332,6 +335,15 @@ version_t LogClient::queue(LogEntry &entry)
   return entry.seq;
 }
 
+void LogClient::reset()
+{
+  std::lock_guard l(log_lock);
+  if (log_queue.size()) {
```
Comment (tchaikov) [inline src/common/LogClient.cc:341]: nit, might want to use `!log_queue.empty()`, but I think it's fine if we just call `log_queue.clear()` without checking if `log_queue` is empty. as this is not a critical path, and the overhead of calling `clear()` is relatively low.

**Review by tchaikov — APPROVED** [review]
LGTM, just a couple nits.

**Review by tchaikov — CHANGES_REQUESTED** [review]
oh, the title of "osd: fix assertion when disabling and re-enabling clog_to_monitors" needs more work:

- might want to replace "osd" with "common"
- might want to be more specific in the title, like "reset last_log_sent when log_to_monitors is updated". and put the details in the commit message. you could copy part of the comment in https://github.com/ceph/ceph/pull/38997#issue-558980766 in the commit message.

**Comment by gerald-yang** [comment]
Thanks for the review, I will update the PR soon

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-03-06_09:19:54-rados-wip-kefu-testing-2021-03-06-1553-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-03-07_12:30:50-rados-wip-kefu-testing-2021-03-07-1842-distro-basic-smithi/


failures tracked by

- https://tracker.ceph.com/issues/49638
- https://tracker.ceph.com/issues/49484

**Pull request merged**